### PR TITLE
!feat: Make SpokePools upgradeable 

### PR DESCRIPTION
**Pull request by nicholaspai**
This PR makes SpokePools compatible with OpenZeppelin's upgradeable proxy pattern. Some important changes to support upgradeability:

- `SpokePool's constructor` adds the `_disableInitializers` call to [prevent it from being left uninitialized](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable#initializing_the_implementation_contract).
- All non-constant state variables that have initial values are [instead set in the `initialize` function](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable#initializing_the_implementation_contract).
- Contracts that the `SpokePool` extends such as `Lockable` and `Testable` that have constructors replaces those with upgradeable versions that replace constructors with initializers.
- All unit tests involving the spoke pool are changed to interface with an upgradeable version of the spoke pool

### CODE REVIEW

File: contracts/upgradeable/LockableUpgradeable.sol [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: AGPL-3.0-only
```
Comment (mrice32) [inline contracts/upgradeable/LockableUpgradeable.sol:1]: As discussed IRL, maybe we should just use the OZ lockable contract? Do you see any problem with that?

File: contracts/upgradeable/MultiCallerUpgradeable.sol [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+/**
+ * @title MockSpokePool
+ * @notice Logic is 100% copied from "@uma/core/contracts/common/implementation/MultiCaller.sol" but one
```
Comment (mrice32) [inline contracts/upgradeable/MultiCallerUpgradeable.sol:6]: We should remember to exclude this contract from the audit scope considering the contract's logic has already been audited.

**Review by mrice32 — APPROVED** [review]
LGTM!

File: deployments/deployments.json [hunk 3]
```diff
@@ -2,7 +2,8 @@
   "1": {
     "AcrossConfigStore": { "address": "0x3B03509645713718B78951126E0A6de6f10043f5", "blockNumber": 14717196 },
     "AcrossMerkleDistributor": { "address": "0xE50b2cEAC4f60E840Ae513924033E753e2366487", "blockNumber": 15976846 },
-    "Arbitrum_Adapter": { "address": "0x937958992799bF4A9a656E6596FD10d7Da5c2216", "blockNumber": 14704380 },
+    "Arbitrum_Adapter": { "address": "0x29528780E29abb8Af95a5e5a125b94766987543F", "blockNumber": 16243583 },
+    "Arbitrum_RescueAdapter": { "address": "0xC6fA0a4EBd802c01157d6E7fB1bbd2ae196ae375", "blockNumber": 16233939 },
```
Comment (pxrl) [inline deployments/deployments.json:6]: Seems unrelated - maybe split it out?

File: test/chain-specific-spokepools/Arbitrum_SpokePool.ts [hunk 4]
```diff
@@ -23,14 +23,28 @@ describe("Arbitrum Spoke Pool", function () {
...
 
+  it("Only cross domain owner upgrade logic contract", async function () {
+    // TODO: Could also use upgrades.prepareUpgrade but I'm unclear of differences
+    const implementation = await hre.upgrades.deployImplementation(
+      await getContractFactory("Arbitrum_SpokePool", owner),
+      { kind: "uups" }
+    );
+
+    // upgradeTo fails unless called by cross domain admin
+    await expect(arbitrumSpokePool.upgradeTo(implementation)).to.be.reverted;
```
Comment (pxrl) [inline test/chain-specific-spokepools/Arbitrum_SpokePool.ts:44]: Is it worthwhile specifying the expected revert message for these? Then it's clear exactly where it's being caught.

File: test/gas-analytics/SpokePool.RelayerRefundLeafExecution.ts [hunk 5]
```diff
@@ -133,7 +133,7 @@ describe("Gas Analytics: SpokePool Relayer Refund Leaf Execution", function () {
       );
 
       // Store new tree.
-      await spokePool.connect(dataWorker).relayRootBundle(
+      await spokePool.connect(owner).relayRootBundle(
```
Comment (pxrl) [inline test/gas-analytics/SpokePool.RelayerRefundLeafExecution.ts:136]: Are these related to SpokePool upgradability?

File: contracts/SpokePool.sol [hunk 6]
```diff
@@ -129,23 +135,37 @@ abstract contract SpokePool is SpokePoolInterface, Testable, Lockable, MultiCall
     event PausedDeposits(bool isPaused);
     event PausedFills(bool isPaused);
 
+    /**
+     * Do not leave an implementation contract uninitialized. An uninitialized implementation contract can be
+     * taken over by an attacker, which may impact the proxy. To prevent the implementation contract from being
+     * used, you should invoke the _disableInitializers function in the constructor to automatically lock it when
+     * it is deployed: */
```
Comment (pxrl) [inline contracts/SpokePool.sol:142]:
```suggestion
     * it is deployed:
     */
```

File: contracts/SpokePool.sol [hunk 7]
```diff
@@ -38,11 +44,11 @@ abstract contract SpokePool is SpokePoolInterface, Testable, Lockable, MultiCall
 
     // Address of wrappedNativeToken contract for this network. If an origin token matches this, then the caller can
     // optionally instruct this contract to wrap native tokens when depositing (ie ETH->WETH or MATIC->WMATIC).
-    WETH9 public immutable wrappedNativeToken;
+    WETH9 public wrappedNativeToken;
 
     // Any deposit quote times greater than or less than this value to the current contract time is blocked. Forces
     // caller to use an approximately "current" realized fee. Defaults to 10 minutes.
-    uint32 public depositQuoteTimeBuffer = 600;
+    uint32 public depositQuoteTimeBuffer;
```
Comment (pxrl) [inline contracts/SpokePool.sol:51]: It [won't be possible to add any more state variables](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable#modifying-your-contracts) to the base SpokePool after deployment. Have we ruled out adding some placeholders, just in case we find a use for something else in future?

File: deployments/deployments.json (hunk 3 above)
Comment (nicholaspai) [reply deployments/deployments.json:6]: This was accidentally removed so i'm just adding it back here

File: test/gas-analytics/SpokePool.RelayerRefundLeafExecution.ts (hunk 5 above)
Comment (nicholaspai) [reply test/gas-analytics/SpokePool.RelayerRefundLeafExecution.ts:136]: they're not, just different test fixes

File: contracts/upgradeable/LockableUpgradeable.sol (hunk 1 above)
Comment (nicholaspai) [reply contracts/upgradeable/LockableUpgradeable.sol:1]: Nope, I looked into Lockable. The main feature we add is the `_startReentrantGuardDisabled` internal function but that's not used by Across contracts so I think we're safe to use OZ lockable

File: contracts/SpokePool.sol (hunk 7 above)
Comment (nicholaspai) [reply contracts/SpokePool.sol:51]: Great point https://github.com/across-protocol/contracts-v2/pull/218/commits/6a9f5093193229800f4e46aaa52dd1d54e3fe667

What do you and @mrice32 think about adding 50 storage slots for future upgrades?

Comment (mrice32) [reply contracts/SpokePool.sol:51]: Great call! Any reason not to make it way way bigger than we'd ever need? Like 1000? Just to be super safe

File: contracts/upgradeable/LockableUpgradeable.sol (hunk 1 above)
Comment (mrice32) [reply contracts/upgradeable/LockableUpgradeable.sol]: Awesome!

**Review by mrice32 — APPROVED** [review]
LGTM!

File: contracts/SpokePool.sol (hunk 7 above)
Comment (nicholaspai) [reply contracts/SpokePool.sol:51]: Here's a thought experiment. Imagine we wanted to support bridging with arbitrary data so that someone could bridge-and-swap. Say we had to upgrade the `deposit()` function (or add a new depositWithData function). Would this be possible with upgradeable proxies? How much of the extra storage slots would we need?

Comment (mrice32) [reply contracts/SpokePool.sol:51]: It depends on the implementation. If the data is just hashed into the deposit data and added to an event, no storage changes would be necessary.

If we needed to store extra info about each deposit, deposits would need extra storage, but they aren't stored continuously, so that can be done without pre-reserving slots.

In this implementation, what extra contiguous storage would you imagine needing?

Comment (nicholaspai) [reply contracts/SpokePool.sol:51]: I'm not sure, this was just a functionality that I discussed with Ryan about adding and wanted to make sure we're prepared for it. Basically bridge-and-swap would be a good UX addition

Comment (nicholaspai) [reply contracts/SpokePool.sol:51]: What if we needed to add a new Base contract function, does that take up slots?

Comment (mrice32) [reply contracts/SpokePool.sol:51]: Nope, functions don't take up slots or anything. They are just more bytecode in the implementation contract.

**Review by mrice32 — COMMENTED** [review]
https://github.com/across-protocol/contracts-v2/pull/218#discussion_r1093504016

File: deployments/deployments.json (hunk 3 above)
Comment (pxrl) [reply deployments/deployments.json:6]: It's preferable from a commit hygiene standpoint to split it out and submit it separately - at least so it's not ++confusing when someone looks back at the commit history and sees it modified as part of something totally unrelated. Just my 2c.

**Review by pxrl — APPROVED** [review]
Nice!

**Comment by pxrl** [comment]
@nicholaspai One final thought: the OVM_SpokePool is also a base contract. Not sure whether we want to reserve some slots there as well?

**Comment by nicholaspai** [comment]
> @nicholaspai One final thought: the OVM_SpokePool is also a base contract. Not sure whether we want to reserve some slots there as well?

Great call, i've added 50 slots for OVM_SpokePool and updated SpokePool's __gap to 1000

cc @mrice32

**Review by mrice32 — COMMENTED** [review]
> > @nicholaspai One final thought: the OVM_SpokePool is also a base contract. Not sure whether we want to reserve some slots there as well?
> 
> Great call, i've added 100 slots for OVM_SpokePool and updated SpokePool's __gap to 1000
> 
> cc @mrice32

@nicholaspai why not just use a 1000 gap everywhere to give us tons of room and standardize? We have 2^256 slots to work with here, so there's no running out :).

**Review by mrice32 — APPROVED** [review]
LGTM!

**Comment by nicholaspai** [comment]
@dohaki @mrice32 @pxrl I had to update this to take into account the new EIP712 contract that the SpokePool extends. This required changing that contract to be upgradeable with an initializer function instead of a constructor see this commit https://github.com/across-protocol/contracts-v2/pull/218/commits/22f5fbed937fa1be2d7e5bcd8bd8dbb834172cfa#diff-8c305f0568345cd237252c531582ce424adc9c7c9151cfefe304fd389b874c59 !

File: contracts/upgradeable/EIP712CrossChainUpgradeable.sol [hunk 8]
```diff
@@ -72,6 +75,13 @@ abstract contract EIP712CrossChain {
...
-        return ECDSA.toTypedDataHash(_domainSeparatorV4(originChainId), structHash);
+        return ECDSAUpgradeable.toTypedDataHash(_domainSeparatorV4(originChainId), structHash);
     }
+
+    /**
+     * @dev This empty reserved space is put in place to allow future versions to add new
+     * variables without shifting down storage in the inheritance chain.
+     * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
+     */
+    uint256[1000] private __gap;
```
Comment (mrice32) [inline contracts/upgradeable/EIP712CrossChainUpgradeable.sol:86]: Why is the gap at the bottom of the file below the functions? Shouldn't it be at the end of the list of variables, like in the other contracts?

File: contracts/upgradeable/MultiCallerUpgradeable.sol [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+            results[i] = result;
+        }
+    }
+
+    /**
+     * @dev This empty reserved space is put in place to allow future versions to add new
+     * variables without shifting down storage in the inheritance chain.
+     * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
+     */
+    uint256[1000] private __gap;
```
Comment (mrice32) [inline contracts/upgradeable/MultiCallerUpgradeable.sol:41]: Same as above.

File: contracts/upgradeable/TestableUpgradeable.sol [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            return block.timestamp; // solhint-disable-line not-rely-on-time
+        }
+    }
+
+    /**
+     * @dev This empty reserved space is put in place to allow future versions to add new
+     * variables without shifting down storage in the inheritance chain.
+     * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
+     */
+    uint256[1000] private __gap;
```
Comment (mrice32) [inline contracts/upgradeable/TestableUpgradeable.sol:60]: Same as above.

**Review by mrice32 — APPROVED** [review]
LGTM!

File: contracts/upgradeable/EIP712CrossChainUpgradeable.sol (hunk 8 above)
Comment (nicholaspai) [reply contracts/upgradeable/EIP712CrossChainUpgradeable.sol:86]: I'm not sure this is how OZ adds their gaps so I was following along https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/proxy/utils/UUPSUpgradeable.sol#L111

**Review by dohaki — APPROVED** [review]
Awesome LGTM 👍 I mainly looked at the changes related to EIP712

File: contracts/upgradeable/EIP712CrossChainUpgradeable.sol (hunk 8 above)
Comment (mrice32) [reply contracts/upgradeable/EIP712CrossChainUpgradeable.sol:86]: Makes sense. Either way is totally fine! However, I think this PR has it done both ways. Ovm_SpokePool and SpokePool seem to have it at the top. As long as it's consistent, LGTM!

**Review by mrice32 — COMMENTED** [review]
https://github.com/across-protocol/contracts-v2/pull/218#discussion_r1097529817

File: contracts/upgradeable/EIP712CrossChainUpgradeable.sol [hunk 11]
```diff
@@ -1,27 +1,35 @@
...
-import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts-upgradeable/utils/cryptography/ECDSAUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
 
 /**
  * @dev https://eips.ethereum.org/EIPS/eip-712[EIP 712] is a standard for hashing and signing of typed structured data.
  *
  * This contract is based on OpenZeppelin's implementation:
- * https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.8.0/contracts/utils/cryptography/EIP712.sol
+ * https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/utils/cryptography/EIP712Upgradeable.sol
```
Comment (pxrl) [inline contracts/upgradeable/EIP712CrossChainUpgradeable.sol:11]:
```suggestion
 * https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/release-v4.8/contracts/utils/cryptography/EIP712Upgradeable.sol
```

Referring to master could be misleading in future, so it probably makes sense to refer to a specific version. I expect that we're just migrating from openzeppelin-contracts v4.8.0 to openzeppelin-contracts-upgradeable v4.8.0 - is that correct?

File: contracts/upgradeable/EIP712CrossChainUpgradeable.sol [hunk 12]
```diff
@@ -1,27 +1,35 @@
...
  */
-abstract contract EIP712CrossChain {
+abstract contract EIP712CrossChainUpgradeable is Initializable {
     /* solhint-disable var-name-mixedcase */
-    bytes32 private immutable _HASHED_NAME;
-    bytes32 private immutable _HASHED_VERSION;
-    bytes32 private immutable _TYPE_HASH;
+    bytes32 private _HASHED_NAME;
+    bytes32 private _HASHED_VERSION;
+    bytes32 private constant _TYPE_HASH = keccak256("EIP712Domain(string name,string version,uint256 chainId)");
```
Comment (pxrl) [inline contracts/upgradeable/EIP712CrossChainUpgradeable.sol:22]: Should this be assigned in initialize()?

https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable#avoid-initial-values-in-field-declarations

**Review by pxrl — APPROVED** [review]
Would be nice to be more specific in the comment referring to the OZ EIP712Upgradable contract, but otherwise :+1:
